jv-oop-advanced

noleynik29 opened the pull request:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).

 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+    private ColorSupplier colorSupplier = new ColorSupplier();
+
+    public Figure getRandomFigure() {
+        int type = random.nextInt(5); // 0..4
```
This violates checklist item: "All magic numbers in your code should be constants." The value 5 (number of figure types) should be a named constant.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+        String color = colorSupplier.getRandomColor();
+
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don't pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don't pass expressions in the constructor." Assign the random values to local variables before passing them to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don't pass expressions in the constructor." Assign the random values to local variables before passing them to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 3 -> new Circle(color, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don't pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 3 -> new Circle(color, random.nextInt(10) + 1);
+            default -> new IsoscelesTrapezoid(color,
+                    random.nextInt(10) + 1, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don't pass expressions in the constructor." Assign the random values to local variables before passing them to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle("white", 10);
```
This violates checklist item: "All magic numbers in your code should be constants." The value 10 (radius for default circle) should be a named constant.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Figure.java:
```diff
@@ -0,0 +1,12 @@
+package core.basesyntax.model;
+
+public abstract class Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes." You must use interfaces to define behavior (getArea, draw) instead of abstract methods in an abstract class.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because there are critical checklist violations that must be fixed before approval.

Critical issues to fix:
1. Use interfaces for behavior, not an abstract class
   - Files: src/main/java/core/basesyntax/model/Figure.java and all figure subclasses
   - Why: The checklist requires using interfaces to define behavior (e.g., getArea, draw). Using an abstract class with abstract methods to enforce behavior violates this rule.
   - How: Create interfaces (e.g., `AreaSupplier` with `getArea()`, `Drawable` with `draw()`), implement them in each figure class, and make `Figure` a concrete class holding shared state (like `color`) without abstract methods. Alternatively, remove `Figure` and have each figure implement the interfaces and manage its own color field.
   - Where: In Figure.java, remove abstract methods and the abstract modifier; in Circle/Rectangle/Square/RightTriangle/IsoscelesTrapezoid, update `extends Figure` usage accordingly and implement the interfaces.

2. Replace magic numbers with named constants
   - File: src/main/java/core/basesyntax/FigureSupplier.java
   - Why: The checklist states all magic numbers must be constants. Values like 5 (figure types), 10 (random upper bound / default radius), and 1 (offset) should be constants.
   - How: Introduce `private static final` constants, e.g., `FIGURE_TYPES_COUNT`, `RANDOM_BOUND`, `RANDOM_MIN`, `DEFAULT_RADIUS`, etc., and use them instead of literals.
   - Where: In getRandomFigure() switch range and random generation, and in getDefaultFigure() for the default circle radius.

3. Don’t pass expressions directly into constructors when creating figures
   - File: src/main/java/core/basesyntax/FigureSupplier.java
   - Why: The checklist requires assigning random values to local variables before passing them to constructors for readability and maintainability.
   - How: For each case in `getRandomFigure()`, first assign random dimensions to local variables (e.g., `int radius = random.nextInt(...) + ...;`) and then pass those variables to the constructor.
   - Where: Each constructor call in the switch cases. Example: for Circle, compute `radius` first, then call `new Circle(color, radius)`.

Minor notes:
- Your functional logic, area calculations, and output content look good. ColorSupplier and main flow in HelloWorld are well-structured.

Once you address the above items, your solution should fully comply. Keep up the good work—your core implementation is solid; it just needs these compliance refactors.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+    private ColorSupplier colorSupplier = new ColorSupplier();
+
+    public Figure getRandomFigure() {
+        int type = random.nextInt(5); // 0..4
```
This violates checklist item: "All magic numbers in your code should be constants." Replace 5, 10, and 1 with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+        String color = colorSupplier.getRandomColor();
+
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign random values to local variables before passing them to constructors.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign random values to local variables before passing them to constructors.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+        return switch (type) {
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign random values to local variables before passing them to constructors.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+            case 0 -> new Square(color, random.nextInt(10) + 1);
+            case 1 -> new Rectangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 3 -> new Circle(color, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign random values to local variables before passing them to constructors.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+            case 2 -> new RightTriangle(color, random.nextInt(10) + 1, random.nextInt(10) + 1);
+            case 3 -> new Circle(color, random.nextInt(10) + 1);
+            default -> new IsoscelesTrapezoid(color,
+                    random.nextInt(10) + 1, random.nextInt(10) + 1, random.nextInt(10) + 1);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign random values to local variables before passing them to constructors.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle("white", 10);
```
This violates checklist item: "All magic numbers in your code should be constants." Use a constant for the default color ("white") and default radius (10).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Circle.java:
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.model;
+
+public class Circle extends Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." Circle should implement interfaces (e.g., AreaSupplier, Drawable) directly, not extend an abstract Figure class that enforces behavior.

mateacademy-ai-mentor replied:
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." The Circle class should not extend Figure. Instead, it should implement Drawable and FigureArea interfaces directly and store color as a field.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Figure.java:
```diff
@@ -0,0 +1,13 @@
+
+import core.basesyntax.Drawable;
+
+public abstract class Figure implements Drawable {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." Use interfaces for getArea() and draw() instead of an abstract class.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Figure.java:
```diff
@@ -0,0 +1,13 @@
+        this.color = color;
+    }
+
+    public abstract double getArea();
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." Use interfaces for getArea() and draw() instead of an abstract class.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/IsoscelesTrapezoid.java:
```diff
@@ -0,0 +1,29 @@
+package core.basesyntax.model;
+
+public class IsoscelesTrapezoid extends Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." IsoscelesTrapezoid should implement interfaces (e.g., AreaSupplier, Drawable) directly, not extend an abstract Figure class that enforces behavior.

mateacademy-ai-mentor replied:
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." The IsoscelesTrapezoid class should not extend Figure. Instead, it should implement Drawable and FigureArea interfaces directly and store color as a field.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Rectangle.java:
```diff
@@ -0,0 +1,24 @@
+package core.basesyntax.model;
+
+public class Rectangle extends Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." Rectangle should implement interfaces (e.g., AreaSupplier, Drawable) directly, not extend an abstract Figure class that enforces behavior.

mateacademy-ai-mentor replied:
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." The Rectangle class should not extend Figure. Instead, it should implement Drawable and FigureArea interfaces directly and store color as a field.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/RightTriangle.java:
```diff
@@ -0,0 +1,24 @@
+package core.basesyntax.model;
+
+public class RightTriangle extends Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." RightTriangle should implement interfaces (e.g., AreaSupplier, Drawable) directly, not extend an abstract Figure class that enforces behavior.

mateacademy-ai-mentor replied:
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." The RightTriangle class should not extend Figure. Instead, it should implement Drawable and FigureArea interfaces directly and store color as a field.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Square.java:
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax.model;
+
+public class Square extends Figure {
```
This violates the checklist item: "Don’t use abstract classes to set behavior for classes where interfaces are appropriate for different concerns in this task." Square should implement interfaces (e.g., AreaSupplier, Drawable) directly, not extend an abstract Figure class that enforces behavior.

mateacademy-ai-mentor replied:
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." The Square class should not extend Figure. Instead, it should implement Drawable and FigureArea interfaces directly and store color as a field.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your solution is very close functionally and reads well, but there are a few checklist violations that are considered critical blockers for approval.

**Critical issues to fix (blockers):**
1. Use of abstract class to enforce behavior
   - Files: `src/main/java/core/basesyntax/model/Figure.java`, and all subclasses (`Circle`, `Square`, `Rectangle`, `RightTriangle`, `IsoscelesTrapezoid`).
   - Why: The checklist states you must not use an abstract class to set behavior where interfaces are appropriate. Behaviors `getArea()` and `draw()` should be defined via interfaces (e.g., `AreaSupplier`, `Drawable`).
   - How to fix:
     - Remove behavior enforcement from `Figure`. Either delete `Figure` or make it a simple concrete holder (e.g., `color`) without abstract methods.
     - Ensure each figure class implements the right interfaces directly (you already have `Drawable`; add an `AreaSupplier` interface with `double getArea()` and implement it in each figure).
     - Move any shared color handling to each class (or a small concrete parent without abstract behavior).
     - Example: `class Circle implements Drawable, AreaSupplier { private final String color; ... public double getArea() { ... } public void draw() { ... } }`

2. Magic numbers and constructor expressions in `FigureSupplier`
   - File: `src/main/java/core/basesyntax/FigureSupplier.java`
   - Why: Checklist items require replacing magic numbers with named constants and forbid passing expressions directly into constructors.
   - What to change:
     - Replace magic numbers with constants:
       - Number of figure types (`5`)
       - Random upper bound (`10`) and offset (`1`)
       - Default color (`"white"`) and default radius (`10`)
       - Suggested: `private static final int FIGURE_TYPES = 5;`, `private static final int RANDOM_UPPER_BOUND = 10;`, `private static final int RANDOM_OFFSET = 1;`, `private static final String DEFAULT_COLOR = "white";`, `private static final int DEFAULT_RADIUS = 10;`
     - Don’t pass expressions to constructors. Assign to locals first:
       - Instead of `new Circle(random.nextInt(10) + 1, color)`, do:
         - `int radius = random.nextInt(RANDOM_UPPER_BOUND) + RANDOM_OFFSET;`
         - `Circle circle = new Circle(radius, color);`
     - Apply similarly for width/height/legs/bases in other figures and in `getDefaultFigure()`.

**Minor notes (nice-to-have after blockers):**
- None beyond the above; overall structure, English-only output, and main flow look good.

If you address the above, your solution should be fully compliant. You’ve done a solid job on functionality and clarity—refactoring to interfaces and constants will align it perfectly with the checklist. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+
+        return switch (type) {
+            case 0 -> {
+                int side = random.nextInt(10) + 1;
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The value 10 (upper bound for random) and 1 (offset) should be replaced with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+        return switch (type) {
+            case 0 -> {
+                int side = random.nextInt(10) + 1;
+                yield new Square(color, side);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+                yield new Square(color, side);
+            }
+            case 1 -> {
+                int width = random.nextInt(10) + 1;
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The value 10 (upper bound for random) and 1 (offset) should be replaced with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+            case 1 -> {
+                int width = random.nextInt(10) + 1;
+                int height = random.nextInt(10) + 1;
+                yield new Rectangle(color, width, height);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+                yield new Rectangle(color, width, height);
+            }
+            case 2 -> {
+                int firstLeg = random.nextInt(10) + 1;
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The value 10 (upper bound for random) and 1 (offset) should be replaced with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+            case 2 -> {
+                int firstLeg = random.nextInt(10) + 1;
+                int secondLeg = random.nextInt(10) + 1;
+                yield new RightTriangle(color, firstLeg, secondLeg);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+                yield new RightTriangle(color, firstLeg, secondLeg);
+            }
+            case 3 -> {
+                int radius = random.nextInt(10) + 1;
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The value 10 (upper bound for random) and 1 (offset) should be replaced with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+            }
+            case 3 -> {
+                int radius = random.nextInt(10) + 1;
+                yield new Circle(color, radius);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+                yield new Circle(color, radius);
+            }
+            default -> {
+                int base1 = random.nextInt(10) + 1;
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The value 10 (upper bound for random) and 1 (offset) should be replaced with named constants.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+                int base1 = random.nextInt(10) + 1;
+                int base2 = random.nextInt(10) + 1;
+                int height = random.nextInt(10) + 1;
+                yield new IsoscelesTrapezoid(color, base1, base2, height);
```
This violates checklist item: "Creating a figure, don’t pass expressions in the constructor." Assign the random value to a local variable before passing it to the constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle("white", DEFAULT_CIRCLE_RADIUS);
```
This violates checklist item: "All magic numbers (like 10, 1, 5, etc.) should be constants." The default color string "white" should be a named constant (e.g., DEFAULT_COLOR = "white").

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/Figure.java:
```diff
@@ -0,0 +1,12 @@
+import core.basesyntax.Drawable;
+import core.basesyntax.FigureArea;
+
+public abstract class Figure implements Drawable, FigureArea {
```
This violates checklist item: "Don’t use abstract classes to set behavior (use interfaces instead)." Remove the Figure abstract class and have each figure class implement Drawable and FigureArea directly, storing color as a field.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: Great job overall—your solution is functional, well-structured, and uses interfaces to separate concerns for drawing and area where required. ColorSupplier, Drawable, FigureArea, and HelloWorld look solid.

Critical improvements to address next:
- FigureSupplier: Replace magic numbers with constants and avoid passing expressions directly to constructors.
  - Examples (FigureSupplier.java):
    - Lines where random dimensions use `10` and `+ 1`: replace with constants like `RANDOM_DIMENSION_BOUND` and `RANDOM_DIMENSION_OFFSET`.
    - Default color "white": use a constant like `DEFAULT_COLOR`.
    - Why: This satisfies the checklist and makes the code maintainable and self-documenting.
    - How: Define constants at the top of the class and assign random values to local variables, then pass those variables to constructors (e.g., `int side = random.nextInt(RANDOM_DIMENSION_BOUND) + RANDOM_DIMENSION_OFFSET;` then `new Square(color, side)`).
- Remove abstract class usage for behavior:
  - Files: model/Figure.java and all figure classes (Circle, Square, Rectangle, RightTriangle, IsoscelesTrapezoid).
  - Why: The checklist requires “Don’t use abstract classes to set behavior (use interfaces instead).”
  - How: Delete Figure.java, let each figure `implements Drawable, FigureArea`, and keep `color` as a field within each class. Update constructors accordingly.

What you did well:
- Correct area formulas and clear, English-only draw() output.
- Single FigureSupplier instance and clean main flow.

You’re very close—polish those constants and refactor away the abstract Figure base, and you’ll be in full compliance. Keep up the good work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).

noleynik29 committed:
There are some figures of the following types: square, rectangle, right triangle, circle, isosceles trapezoid. I've to created corresponding classes for them(Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid) All figures have                                                                - state - all figures have color, but each figure type can also have one or several unique properties (radius for circle, firstLeg and secondLeg for right triangle, and so on).
 - behavior - we can obtain the area of any figure and are able to draw it. To 'draw' means to print out all information about a figure using System.out.println() (you shouldn't override the toString() method for this).
